vllm/core/scheduler.py

```diff
@@ -370,9 +546,13 @@ def schedule(self) -> Tuple[List[SequenceGroupMetadata], SchedulerOutputs]:
                 block_tables[seq_id] = self.block_manager.get_block_table(seq)
                 self.block_manager.access_all_blocks_in_seq(seq, now)
 
+            # It assumes the scheduled_seq_groups is ordered by
+            # prefill < decoding.
+            is_prompt = i < scheduler_outputs.num_prompt_groups
```

Comment:
For chunked prefill do we still need to distinguish between prefill/subquery/generation with flags? I feel this is only about how many tokens you have in the context and how many tokens you are actually running the forward pass.

Reply:
I think the scheduler itself should be aware of prefill/chunked/decoding to decide priorities. For worker perspective (SeqGroupMetadata), I agree it doesn't have to be known in theory, but it may require bigger changes. Some examples;

- (IIUC) there are some attention impl that are more optimized for length 1 token (i.e., decoding). For example, our paged attention kernel (IIUC, it is optimized for gemv), or flash decoding https://pytorch.org/blog/flash-decoding/. If we don't expose this info to worker, we need to add code there to detect query size and apply different attention kernels. if you see the e2e PR, we still distinguish prefill/chunked prefill from decoding for this reason (https://github.com/vllm-project/vllm/pull/3483). 
- When preparing inputs, depending on prefill vs chunked prefill, we should prepare inputs differently. For example, we don't prepare block tables for pure prefill VS need to prepare block tables for chunked prefill (AND prefix caching). 

While both of items are doable, I prefer to first go with the current direction (less evading and more iterative) and think about clean up more holistically. I will make sure to own this and clean up. I think it will be much easier to discuss the direction once e2e PRs are merged to master to clean it up then doing it right now. Wdyt?